CRM/Utils/System/Joomla.php

```diff
@@ -507,6 +503,26 @@ public function getVersion() {
     }
   }
 
+  public function getJVersion($joomlaBase) {
+    // Files may be in different places depending on Joomla version
+    if (!defined('JVERSION')) {
+      // Joomla 3.8.0+
+      $versionPhp = $joomlaBase . '/libraries/src/Version.php';
+      if (!file_exists($versionPhp)) {
+        // Joomla < 3.8.0
+        $versionPhp = $joomlaBase . '/libraries/cms/version/version.php';
+      }
+      require $versionPhp;
+      $jversion = new JVersion();
+      define('JVERSION', $jversion->getShortVersion());
+    }
+  }
+
+  public function getBasePath() {
+    // Setup the base path related constant.
+    return dirname(dirname(dirname(dirname(dirname(dirname(dirname(dirname(__FILE__))))))));
```

Comment:
for getting Joomla base path , dirname not work properly when civicrm as soft link
e.g /var/www/html/joomla/administrator/components/com_civicrm/civicrm -> /var/www/html/civicrm
instead we can use $civicrm_root variable

```
global $civicrm_root;
$joomlaPath = explode('/administrator', $civicrm_root);
$joomlaBase = $joomlaPath[0];
```

Reply:
I don't like the dirname either - it was there in 2 places.  All I have done is break it into it's own function so that it can be replaced with something better more easily.  As this is my first time using Joomla with CiviCRM I'm not too familiar with it.
@sunilpawar Are you happy that $civicrm_root will be defined in all cases? Eg. when run from cli.php etc.?  If you can be certain I will add your change to my PR.

Reply:
@mattwire civicrm.settings.php always loaded first, $civicrm_root is defined in that file. so it always present.

Reply:
@sunilpawar Thankyou.  I've updated this on the other PR #11062 I created which also includes these changes - could you comment on the changes in that other PR for me?